### PR TITLE
IO improvements

### PR DESCRIPTION
**Pull request by ramunasd**
* fix unhandled warning in SocketIO on failed connect
* add heartbeat support for socket connections
* move common methods and properties to abstract class
* fix docblocks

### CODE REVIEW

File: PhpAmqpLib/Wire/IO/SocketIO.php [hunk 1]
```diff
@@ -87,15 +74,15 @@ public function connect()
     }
 
     /**
-     * @return resource
+     * @inheritdoc
      */
-    public function getSocket()
+    public function get_socket()
```
Comment (lukebakken) [inline PhpAmqpLib/Wire/IO/SocketIO.php]: This change isn't backwards-compatible. Can we keep the old name for now?

Comment (ramunasd) [reply PhpAmqpLib/Wire/IO/SocketIO.php]: We can, just for sake of compatibility :)

**Comment by ramunasd** [comment]
I will rebase everything after #642

File: PhpAmqpLib/Wire/IO/SocketIO.php (hunk 1 above)
Comment (lukebakken) [reply PhpAmqpLib/Wire/IO/SocketIO.php]: Thanks! Otherwise this PR could only go into version `3.0.0`

**Comment by lukebakken** [comment]
@ramunasd - #642 is merged, thanks again for that fix.

**Comment by ramunasd** [comment]
More duplicated code resolved.
Looks like HHVM 4.0 is out but composer is not compatible and build fails:
```
Fatal error: Uncaught exception 'InvalidArgumentException' with message 'array_pop() expects parameter 1 by reference, but the call was not annotated with '&'' in phar://composer.phar/bin/../src/../src/Composer/Util/Silencer.php:51
```

It is possible to define hhvm version https://docs.travis-ci.com/user/languages/php/#hhvm-versions-are-available-on-trusty-only

Are we gonna support old hhvm versions only?

**Comment by lukebakken** [comment]
> Are we gonna support old hhvm versions only?

I don't think there's a problem with dropping `hhvm` testing if Travis doesn't support it.

**Comment by ramunasd** [comment]
rebased on latest master

**Comment by lukebakken** [comment]
Thank you!

**Comment by ramunasd** [comment]
This lib is getting better and better. Thank You Luke Bakken for maintainership.

**Pull request merged**